File: 20231020_Friday/20231020.go
// Leetcode: 341. Flatten Nested List Iterator.
// https://leetcode.com/problems/flatten-nested-list-iterator
// = = = = = = = = = = = = = 
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = =
// Runtime: 5 ms, faster than 63.08% of Go online submissions for Flatten Nested
// List Iterator.
// Memory Usage: 5.6 MB, less than 30.77% of Go online submissions for Flatten
// Nested List Iterator.
// 2023.10.20 Daily Challenge.

/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * type NestedInteger struct {
 * }
 *
 * // Return true if this NestedInteger holds a single integer, rather than a nested list.
 * func (this NestedInteger) IsInteger() bool {}
 *
 * // Return the single integer that this NestedInteger holds, if it holds a single integer
 * // The result is undefined if this NestedInteger holds a nested list
 * // So before calling this method, you should have a check
 * func (this NestedInteger) GetInteger() int {}
 *
 * // Set this NestedInteger to hold a single integer.
 * func (n *NestedInteger) SetInteger(value int) {}
 *
 * // Set this NestedInteger to hold a nested list and adds a nested integer to it.
 * func (this *NestedInteger) Add(elem NestedInteger) {}
 *
 * // Return the nested list that this NestedInteger holds, if it holds a nested list
 * // The list length is zero if this NestedInteger holds a single integer
 * // You can access NestedInteger's List element directly if you want to modify it
 * func (this NestedInteger) GetList() []*NestedInteger {}
 */
package main

type NestedIterator struct {
	i    int
	data []int
}

type pair struct {
	i int
	l []*NestedInteger
}

func Constructor(nestedList []*NestedInteger) *NestedIterator {
	stack := make([]*pair, 0)
	stack = append(stack, &pair{0, nestedList})
	var dt []int
	for 0 != len(stack) {
		ln := len(stack)
		tmp := stack[ln-1]
		stack = stack[0 : ln-1]
		i, lst := tmp.i, tmp.l
		for i < len(lst) {
			if lst[i].IsInteger() {
				dt = append(dt, lst[i].GetInteger())
				i += 1
			} else {
				stack = append(stack, &pair{i + 1, lst})
				lst = lst[i].GetList()
				i = 0
			}
		}
	}

	return &NestedIterator{0, dt}
}

func (this *NestedIterator) Next() int {
	i := this.i
	this.i += 1
	return this.data[i]
}

func (this *NestedIterator) HasNext() bool {
	return this.i < len(this.data)
}

File: 20231020_Friday/20231020.rb
# Leetcode: 341. Flatten Nested List Iterator.
# https://leetcode.com/problems/flatten-nested-list-iterator
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
#
#class NestedInteger
#    def is_integer()
#        """
#        Return true if this NestedInteger holds a single integer, rather than a nested list.
#        @return {Boolean}
#        """
#
#    def get_integer()
#        """
#        Return the single integer that this NestedInteger holds, if it holds a single integer
#        Return nil if this NestedInteger holds a nested list
#        @return {Integer}
#        """
#
#    def get_list()
#        """
#        Return the nested list that this NestedInteger holds, if it holds a nested list
#        Return nil if this NestedInteger holds a single integer
#        @return {NestedInteger[]}
#        """

class NestedIterator
    # @param {NestedInteger[]} nested_list
    def initialize(nested_list)
      # If it is a flat list need to store only index.
      # If it is a list of list also need a stack. When enter some list need to
      # do something.
      # When current index is equal to size need to go up.
      @l = nested_list # Full list.
      @c = nested_list # Current.
      @s = [] # Stack.
      @i = 0 # Pointer.
        
    end

    # @return {Boolean}
    def has_next
      return true if @c.size > @i
      # Moving down into the stack.
      @s.reverse_each do |pair|
        i,l = pair
        # Found one element in the stack that is not empty.
        return true if i+1 < l.size
      end
      return false
    end

    # @return {Integer}
    def next
      norm
    end
    def norm
      if @i < @c.size && @c[@i].is_integer()
        ans = @c[@i].get_integer()
        @i += 1
        reutrn @ans
      end
      if @i < @c.size && ! @c[@i].is_integer()
        while @i < @c.size && !@c[@i].is_integer()
          @s.push([@i+1,@c])
          @i, @c = 0, @c[@i].get_list()
        end
        ans = @c[@i].get_integer()
        @i += 1
        return ans
      end
      if @i >= @c.size
        @i,@c = @s.pop
        return norm
      end
    end
end

# Your NestedIterator will be called like this:
# i, v = NestedIterator.new(nested_list), []
# while i.has_next()
#    v << i.next
# end
